tests/testthat/test-slice-assign.R

```diff
@@ -83,13 +83,23 @@ test_that("can assign lists", {
 
 test_that("atomics can't be assigned in lists", {
   x <- list(NULL)
-  expect_error(vec_slice(x, 1) <- NA, class = "vctrs_error_incompatible_type")
-  expect_error(vec_assign(x, 1, NA), class = "vctrs_error_incompatible_type")
+  expect_error(vec_slice(x, 1) <- 1, class = "vctrs_error_incompatible_type")
+  expect_error(vec_assign(x, 1, 2), class = "vctrs_error_incompatible_type")
 
   expect_error(vec_slice(x, 1) <- "foo", class = "vctrs_error_incompatible_type")
   expect_error(vec_assign(x, 1, "foo"), class = "vctrs_error_incompatible_type")
 })
 
+# TODO - Revisit this behavior, as it seems inconsistent from our emerging
+# treatment of `NA` as `list(NULL)` and makes internal handling more
+# complicated
+test_that("unspecified() can be assigned in lists, but NA unspecified cannot", {
```

Comment:
Why do we allow assigning unspecified? This seems odd to me, like unspecified has leaked where it shouldn't.

Reply:
I think our general stance is to always allow assigning unspecified.

`vec_ptype2.logical.list()` was special cased to not allow `vec_assign(list(1), 1, NA)`, but `vec_assign(list(1), 1, unspecified(1))` happened to slip through because it isn't caught by that method. This special case complicates the `vec_type2()` C code and is why I need `vec_ptype2_dispatch_unspecified_list()`.

I personally think that `vec_assign(list(1), 1, NA)` should be allowed, and is well defined. `NA` is cast to `list(NULL)`, and then assigned.

So my suggestion would be to:

- Remove `vec_ptype2.logical.list()` entirely, and let `vec_ptype2.logical.default()` be called

- Remove the C level `vec_ptype2_dispatch_unspecified_list()` special case

- Update this test to allow both `unspecified()` and `NA` to be assigned into lists

Reply:
See https://github.com/r-lib/vctrs/pull/812#discussion_r379320081
I'd just not test `unspecified()` and test `NA` instead, otherwise mark the tests as monitoring instead of definitional.